JSQMessagesViewController/Model/JSQAudioMediaItem.m

```diff
@@ -0,0 +1,389 @@
+        if (self.delegate) {
+            if (error && [self.delegate respondsToSelector:@selector(audioMediaItem:didNotChangeCategory:)]) {
+                [self.delegate audioMediaItem:self didNotChangeCategory:error];
+            } else if ([self.delegate respondsToSelector:@selector(audioMediaItem:didChangeOriginalAudioCategory:originalOptions:)]) {
+                [self.delegate audioMediaItem:self didChangeOriginalAudioCategory:category originalOptions:options];
+            }
+        }
+    }
+    
+    if (_audioPlayer.playing) {
+        [_progressTimer invalidate];
+        _progressTimer = nil;
+        
+        _playButton.selected = NO;
+        [_audioPlayer stop];
+    } else {
+        
+        // fade the button from play to pause
+        [UIView transitionWithView:_playButton
+                          duration:.3
+                           options:UIViewAnimationOptionTransitionCrossDissolve
+                        animations:^{
+                            _playButton.selected = YES;
+                        }
+                        completion:nil];
+        
+        // set a timer for updating the elapsed time
+        _progressTimer = [NSTimer scheduledTimerWithTimeInterval:0.1
+                                                          target:self
+                                                        selector:@selector(updateProgress:)
+                                                        userInfo:nil
+                                                         repeats:YES];
+        [_audioPlayer play];
+    }
+}
+
+#pragma mark - AVAudioPlayerDelegate
+
+- (void)audioPlayerDidFinishPlaying:(AVAudioPlayer *)player
+                       successfully:(BOOL)flag {
+    
+    [_progressTimer invalidate];
+    _progressTimer = nil;
+    
+    // set progress to full, then fade back to the default state
+    _progressView.progress = 1;
+    [UIView transitionWithView:self.cachedMediaView
+                      duration:.3
+                       options:UIViewAnimationOptionTransitionCrossDissolve
+                    animations:^{
+                        _progressView.progress = 0;
+                        _playButton.selected = NO;
+                        _progressLabel.text = [self timestampString:_audioPlayer.duration
+                                                        forDuration:_audioPlayer.duration];
+                    }
+                    completion:nil];
+}
+
+#pragma mark - JSQMessageMediaData protocol
+
+- (CGSize)mediaViewDisplaySize
+{
+    // load button images if not already available
+    if (! self.playButtonImage) {
+        self.playButtonImage = [[UIImage jsq_defaultPlayImage] jsq_imageMaskedWithColor:self.tintColor];
+    }
+    if (! self.pauseButtonImage) {
+        self.pauseButtonImage = [[UIImage jsq_defaultPauseImage] jsq_imageMaskedWithColor:self.tintColor];
+    }
```

Comment:
the `DisplaySize` method seems like an odd place to initialize these for the first time...

could this happen in `-mediaView` instead?